crates/bevy_ecs/src/storage/table.rs

```diff
@@ -32,6 +32,37 @@ impl TableId {
     }
 }
 
+/// A opaque newtype for rows in [`Table`]s. Specifies a single row in a specific table.
+///
+/// Values of this type are retreivable from [`Archetype::entity_table_row`] and can be
+/// used alongside [`Archetype::table_id`] to fetch the exact table and row where an
+/// [`Entity`]'s
+///
+/// Values of this type are only valid so long as entities have not moved around.
+/// Adding and removing components from an entity, or despawning it will invalidate
+/// potentially any table row in the table the entity was previously stored in. Users
+/// should *always* fetch the approripate row from the entity's [`Archetype`] before
+/// fetching the entity's components.
+///
+/// [`Archetype`]: crate::archetype::Archetype
+/// [`Archetype::entity_table_row`]: crate::archetype::Archetype::entity_table_row
+/// [`Archetype::table_id`]: crate::archetype::Archetype::table_id
+/// [`Entity`]: crate::entity::Entity
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct TableRow(usize);
+
+impl TableRow {
+    #[inline]
+    pub(crate) const fn new(index: usize) -> Self {
```

Comment:
```suggestion
    pub const fn new(index: usize) -> Self {
```
I don't think there's any harm in allowing users to create a `TableRow`. It ought to be possible anyway using a newly created world and inserting `index` many entities lol

Reply:
Table (and Archetype) are externally read only. Opening up a public constructor lets users make it no longer opaque. Users can do index arithmetic. While they can already use the row for a different table than intended or use it beyond when it's currently valid, I'd rather avoid exposing more improper uses to users.

I didn't want to make it pub(crate) either since it'd also keep us honest about ensuring the rows were only created in Table::allocate and Archetype::allocate, but that's not fully possible with the changes to WorldQuery.

Reply:
I don't know what we gain by making them actually opaque via `Table::allocate` and `Archetype::allocate`. Even if they're publicly constructable and publicly able to get an index out of we still get all the desired wins from not being able to accidentally pass an `ArchetypeRow` or some other `usize` into a function expected a `TableRow`.

Reply:
While it'd deter users (and ourselves) from passing in the wrong type, it'd just immediately encourage unwrapping the inner index, and rewrapping it into the new type. By making it opaque, we close off the venue for that, and force users (and us) to use `Table::allocate`, `Archetype::allocate`, `EntityLocation::archetype_index`, and `Archetype::entity_table_row` as a source of truth.

Reply:
I dont understand how this would encourage immediately unwrapping and rewrapping the `TableRow`/`ArchetypeRow` that would be entirely pointless logic in most cases.

I also do not see the advantage of disallowing arithmetic, if you can yes you would be better of using a pre-existing `TableRow` from somewhere else but why forbid anyone from being able to do this? It does not seem like that big of a deal to me, especially since you would have to call `index()` somewhere and then do `TableRow::new`. That seems like a big enough warning to anyone reading the code that its doing something not standard.